### DISCUSSION TITLE
UX - Easier playback looping

### DISCUSSION THREAD

**Discussion by jounih** [audacity/audacity#869] [Ideas & Feature Requests]
Most DAW/audio software at the moment has a dedicated playback looping toggle button, and a selection for a looping range above the tracks. This has become a common UX pattern - and while Audacity does have playback looping, it's not visible in the UI and quite hard to discover.

I propose adding a visible looping toggle button to the transport bar, and using the range selection at the top of the track to control what area should be looping, if the playback looping button is toggled on. The looping range should also not disappear when you stop playback, as it does now.

This way the playback looping behaviour is in line with other audio/DAW apps, becomes more discoverable and easier to use.

At the moment the playback looping behaviour is also coupled to the waveform selection - it should be possible to do these separately. Selecting a part of a waveform to apply an effect should be possible independently of changing the playback looping range of the whole track - a source of confusion currently.

<img width="1440" alt="looping off@2x" src="https://user-images.githubusercontent.com/10109867/117793867-93ac6080-b244-11eb-98d3-434a32ee2de2.png">

<img width="1440" alt="loop range selection@2x" src="https://user-images.githubusercontent.com/10109867/122544954-336bc400-d025-11eb-8ae3-11aece8a9370.png">

<img width="1440" alt="audio selection@2x" src="https://user-images.githubusercontent.com/10109867/122544930-2cdd4c80-d025-11eb-9dee-760fcff6a2c8.png">

Thoughts?

UPDATE: updated the loop range selection in the attached images to  match the prototype design here - https://vimeo.com/562162689#t=07m33s

**Comment by SteveDaulton** [comment]
I find "Shift + Play" to be a more efficient than having yet another button.
https://manual.audacityteam.org/man/transport_toolbar.html#play

**Reply by jounih** [reply]
efficient it may be, but not discoverable at all! And quite unintuitive for people coming from other audio software. I don't think you should need to read the manual to find out how to loop a portion of the track.

I think we can still keep it as a keyboard shortcut to enable and start loop playback for power users though, no need to get rid of it

**Comment by JamesCrook** [comment]
In this design the loop button is separated from play.  It isn't obvious that the loop button is something that influences or belongs with play.  

One approach to fix that is to rearrange the transport buttons.  The route I took was instead to put the loop-mode toggle button at the end of the play-at-speed toolbar.  That toolbar is already only about playing.  I regard loop-mode as a feature not for complete beginners, and that play-at-speed toolbar is also not for complete beginners.  I'd argue that's a better grouping.

Why not abolish the second play button?  I'm coming to that.

To make that change work, I also made ALL playback operate via play at speed.  So 'play' on the main transport toolbar now is just play-at-speed but starting at x1 speed (which you can then adjust once it has started playing with the play-at-speed slider).  This fixes an oddity in Audacity that you currently can't click normal play and then adjust the speed.

My original intention was to then lose the second play button.  However I found the extra play button was still useful.  I used it to remember your most recently selected non x1 speed.  Also by changing the icon on it to double or treble triangles (for x2 or x3 speed) it acts as an additional indicator of non-default speed.

I'm trying to remove all confusing duplications of buttons (why do we have three microphones and three loudspeakers?), and had thought to remove one play button.  I can live with keeping the second play button if in normal use it is showing one of the alternative speeds of play and so indicating that it does something different.

**Reply by jounih** [reply]
@JamesCrook I think we should abolish the second play-at-speed button, I find it strange too. Will have a design proposal to incorporate the playback speed control somewhere else (perhaps a dropdown from the main play button) soon!

However I think that is a separate issue not related to playback looping.

In terms of the loop button positioning -  the most common pattern is to position it on the right of the transport controls -

Logic:

![image](https://user-images.githubusercontent.com/10109867/117799434-00762980-b24a-11eb-87c3-920214726b5c.png)

Audition:

![image](https://user-images.githubusercontent.com/10109867/117799562-2bf91400-b24a-11eb-862a-43d6f83a0f3d.png)

Soundcloud:

![image](https://user-images.githubusercontent.com/10109867/117800107-b8a3d200-b24a-11eb-8bb0-882b6252722b.png)

I do think playback looping something even beginner users reach for fairly often - so making it a visible control (rather than a hidden keyboard shortcut) would benefit most users.

**Reply by JamesCrook** [reply]
Perhaps the loop-play toggle should only appear when the playback-speed slider is visible too?

![image](https://user-images.githubusercontent.com/1621525/117806374-30c1c600-b252-11eb-9df1-5c13b945e7dc.png)

_Ocenaudio does not have a loop play toggle button._    

On Windows our transport toolbar has buttons twice the size of any other toolbar.  That is correct for 'Play' and 'Record', the two most important buttons in the GUI.  Shouldn't we be relegating less important buttons from the transport toolbar rather than adding new ones to it?  Every button there really should earn its place there.

Audacity has a small gap to set off the record button from the others, partly because there are too many transport toolbar buttons.  That is part of why the loop toggle looks wrong to me when placed next to the record button.  It looks too much as if it is associated with record rather than with play.  So should we lose that set off?

**Comment by Tantacrul** [comment]
I think the visualisation of the loop region itself is quite unclear. The arrow edges make it feel imprecise too. I'd love to see a different visual solution for it.

In particular, the right and left arrows don't make a gigantic amount of sense and don't 'feel' like looping indicators.

**Reply by jounih** [reply]
This is the current Audacity loop region visualisation - I thought reusing it would save development effort. I do have a couple of alternatives too, will post!

**Reply by petersampsonaudacity** [reply]
If we do have a separate button - then surely it's colorway should be green to indicate that it is a form of playback.

**Comment by jounih** [comment]
Updated the loop region visual and added a link to the prototype video
